fabric-message-api-v1/src/main/java/net/fabricmc/fabric/api/message/v1/ChatDecoratorEvent.java

```diff
@@ -0,0 +1,111 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.message.v1;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.network.ChatDecorator;
+import net.minecraft.text.Text;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+/**
+ * A class for registering a {@link ChatDecorator}. Check the chat decorator documentation
+ * for how chat decorators work. Unlike other events, this uses a functional interface that is
+ * provided by the vanilla game.
+ *
+ * <p>This event uses phases to provide better mod compatibilities between mods that add custom
+ * content and styling. Chat decorators with the styling phase will always apply after the ones with
+ * the content phase. When registering the chat decorator, it is recommended to choose one of
+ * the phases from this interface and pass that to the {@link Event#register(Identifier, Object)}
+ * function. If not given, the chat decorator will run in the default phase, which is between
+ * the content phase and the styling phase.
+ *
+ * <p>When implementing a chat decorator, it is <strong>very important that the decorator returns
+ * the same contents between previewing and submission</strong> - otherwise the game discards the
+ * message because it was improperly signed. The most straightforward way of doing this is by
+ * making sure the decorator returns the same result for a given message. Another way to solve this
+ * issue is to cache the decorated message.
+ *
+ * <p>Example of registering a content phase chat decorator:
+ *
+ * <pre><code>
+ * ChatDecoratorEvent.EVENT.register(ChatDecoratorEvent.CONTENT_PHASE, (sender, message) -> {
+ *     // Add smiley face. Has to copy() to get a MutableText with siblings and styles.
+ *     return message.copy().append(" :)");
+ * });
+ * </code></pre>
+ *
+ * <p>Example of registering a styling phase chat decorator:
+ *
+ * <pre><code>
+ * ChatDecoratorEvent.EVENT.register(ChatDecoratorEvent.STYLING_PHASE, (sender, message) -> {
+ *     // Apply orange color to messages sent by server operators
+ *     if (sender != null && sender.server.getPlayerManager().isOperator(sender.getGameProfile())) {
+ *         return CompletableFuture.completedFuture(
+ *             message.copy().styled(style -> style.withColor(0xFFA500)));
+ *     }
+ *     return CompletableFuture.completedFuture(message);
+ * });
+ * </code></pre>
+ */
+public final class ChatDecoratorEvent {
```

Comment:
Since this isn't part of ServerChatEvents and has no similarly identifying name it's implied that it also runs on the client

Reply:
Guess I can rename to `ServerChatDecoratorEvent`?

Reply:
yes, maybe Decorate even (verb)?

Reply:
This just exposes `ChatDecorator`, I think `ServerChatDecoratorEvent` is fine.